processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go

```diff
@@ -71,5 +78,68 @@ func (d *Detector) Detect(ctx context.Context) (pdata.Resource, error) {
 	attr.InsertString(conventions.AttributeHostType, meta.InstanceType)
 	attr.InsertString(conventions.AttributeHostName, hostname)
 
+	tags, err := connectAndFetchEc2Tags(meta.Region, meta.InstanceID, d.cfg)
+	if err != nil {
+		return res, fmt.Errorf("failed fetching ec2 instance tags: %w", err)
+	}
+
+	for key, val := range tags {
+		attr.InsertString(tagPrefix+key, val)
+	}
+
 	return res, nil
 }
+
+func connectAndFetchEc2Tags(region string, instanceID string, cfg Config) (map[string]string, error) {
+	if len(cfg.Tags) == 0 {
+		return nil, nil
+	}
+
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(region)},
+	)
+	if err != nil {
+		return nil, err
+	}
+	e := ec2.New(sess)
+
+	return fetchEC2Tags(e, instanceID, cfg)
+}
+
+func fetchEC2Tags(svc ec2iface.EC2API, instanceID string, cfg Config) (map[string]string, error) {
+	ec2Tags, err := svc.DescribeTags(&ec2.DescribeTagsInput{
+		Filters: []*ec2.Filter{{
+			Name: aws.String("resource-id"),
+			Values: []*string{
+				aws.String(instanceID),
+			},
+		}},
+	})
+	if err != nil {
+		return nil, err
+	}
+	tags := make(map[string]string)
+	for _, tag := range ec2Tags.Tags {
+		matched, err := regexArrayMatch(cfg.Tags, *tag.Key)
+		if err != nil {
+			return nil, err
+		}
+		if matched {
+			tags[*tag.Key] = *tag.Value
+		}
+	}
+	return tags, nil
+}
+
+func regexArrayMatch(arr []string, val string) (bool, error) {
+	for _, elem := range arr {
+		matched, err := regexp.MatchString(elem, val)
+		if err != nil {
+			return false, err
+		}
```

Comment:
I would rather compile the regexes and fail earlier (in `NewDetector`) instead of here (which would make it fail during the `Detect` call). If we want to `Detect` more than once at some point compiling earlier makes more sense. Doing it this way also means we don't need to return an error from `regexArrayMatch` anymore.

Reply:
Done 👍